exporter/opensearchexporter/sso_trace_exporter.go

```diff
@@ -0,0 +1,94 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package opensearchexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/opensearchexporter"
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+const (
+	// defaultNamespace value is used as SSOTracesExporter.Namespace when component.Config.Namespace is not set.
+	defaultNamespace = "namespace"
+
+	// defaultDataset value is used as SSOTracesExporter.Dataset when component.Config.Dataset is not set.
+	defaultDataset = "default"
+)
+
+type SSOTracesExporter struct {
+	client       *opensearch.Client
+	Namespace    string
+	Dataset      string
+	httpSettings confighttp.HTTPClientSettings
+	telemetry    component.TelemetrySettings
+}
+
+func newSSOTracesExporter(cfg *Config, set exporter.CreateSettings) (*SSOTracesExporter, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &SSOTracesExporter{
+		telemetry:    set.TelemetrySettings,
+		Namespace:    defaultIfEmpty(cfg.Namespace, defaultNamespace),
+		Dataset:      defaultIfEmpty(cfg.Dataset, defaultDataset),
+		httpSettings: cfg.HTTPClientSettings,
+	}, nil
+}
+
+func (s *SSOTracesExporter) Start(_ context.Context, host component.Host) error {
+	httpClient, err := s.httpSettings.ToClient(host, s.telemetry)
+	if err != nil {
+		return err
+	}
+
+	client, err := newOpenSearchClient(s.httpSettings.Endpoint, httpClient, s.telemetry.Logger)
+	if err != nil {
+		return err
+	}
+
+	s.client = client
+	return nil
+}
+
+func (s *SSOTracesExporter) pushTraceData(ctx context.Context, td ptrace.Traces) error {
+	indexer := newTraceBulkIndexer(s.Dataset, s.Namespace)
```

Comment:
This is the only way to make it synchronous. `Close` needs to be called to flush its document channel.

Does the collector design expects each `pushTraceData` call to block until all `td`s are processed?

Bulk indexer is fundamentally non-blocking. I kept it for OpenSearch-specific bits, but it seems to repeatedly clash with how exporters should be ideally implemented.

Would it make more sense to use the bulk REST API directly?

Reply:
I'm okay with this implementation for now.  It's reasonably well encapsulated and the configuration interface is already aligned with other collector components, so we should be able to change course here without significant user impact should this implementation prove problematic.

> Does the collector design expects each pushTraceData call to block until all tds are processed?

It's a bit complicated.  I think the answer is "yes, but" that "processed" is somewhat nebulous.  The expectation is that the caller shouldn't need to concern itself with the provided telemetry once the call returns, but that could be because the telemetry was batched (as in the batch processor) or persisted to a WAL or successfully exported.  I think generally if an exporter can be blocking it's probably best for it to be blocking, though it does still need to support concurrent invocation as multiple pipelines may feed into it and it may be fed by components that produce telemetry concurrently.